driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -102,120 +83,99 @@ class AsyncCommandBatchCursor<T> implements AsyncAggregateResponseBatchCursor<T>
             @Nullable final BsonValue comment,
             final AsyncConnectionSource connectionSource,
             final AsyncConnection connection) {
-        isTrueArgument("maxTimeMS >= 0", maxTimeMS >= 0);
-        this.cursor = new AtomicReference<>();
-        this.initialCommandCursorResult = initFromCommandCursorDocument(connection.getDescription().getServerAddress(),
-                FIRST_BATCH, commandCursorDocument);
-        this.namespace = initialCommandCursorResult.getNamespace();
+        ConnectionDescription connectionDescription = connection.getDescription();
+        this.commandCursorResult = toCommandCursorResult(connectionDescription.getServerAddress(), FIRST_BATCH, commandCursorDocument);
+        this.namespace = commandCursorResult.getNamespace();
         this.limit = limit;
         this.batchSize = batchSize;
         this.maxTimeMS = maxTimeMS;
-        this.decoder = decoder;
+        this.decoder = notNull("decoder", decoder);
         this.comment = comment;
-        this.maxWireVersion = connection.getDescription().getMaxWireVersion();
-        this.firstBatchEmpty = initialCommandCursorResult.getResults().isEmpty();
-
-        this.operationTime = initialCommandCursorResult.getOperationTime();
-        if (cursor.get() != null) {
-            this.connectionSource = notNull("connectionSource", connectionSource).retain();
-            assertNotNull(connection);
-            if (limitReached()) {
-                killCursor(connection);
-            } else {
-                if (connectionSource.getServerDescription().getType() == ServerType.LOAD_BALANCER) {
-                    this.pinnedConnection = connection.retain();
-                    this.pinnedConnection.markAsPinned(Connection.PinningMode.CURSOR);
-                }
-            }
+        this.maxWireVersion = connectionDescription.getMaxWireVersion();
+        this.firstBatchEmpty = commandCursorResult.getResults().isEmpty();
+
+        AsyncConnection connectionToPin = null;
+        boolean releaseServerAndResources = false;
+        if (limitReached()) {
```

Comment:
If simplifying this code helps, it's safe to remove all the `limitReached` checks.  The limit checks are a remnant of when the driver supported OP_QUERY as well as the `find` command for implementing the find helper.   OP_QUERY required the client to enforce the limit, while with the `find` command the server enforces the limit and when it is reached will return the last batch with a cursorId of 0.

Reply:
Ack - will remove and solely rely on the server

Reply:
The batchSize calculation for getMore is also unnecessary, but there are spec tests that still require it.

Reply:
> it's safe to remove all the limitReached checks

Is this also true for the sync cursor?

Reply:
Yes, it's a property of the find command itself, not sync/async distinction

Reply:
Done.

Reply:
> The batchSize calculation for getMore is also unnecessary, but there are spec tests that still require it.

With the latest changes, the `A successful find event with a getmore and the server kills the cursor (<= 4.4)` fails due to the change in how batchSize is set for getMore.  You'll either have to roll back the part of the change that sets the batchSize, or change the test (this is the subject of DRIVERS-1448).

Reply:
Thank you - updated the test(s)